src/simperium/channel.js

```diff
@@ -104,40 +104,38 @@ internal.removeAndSend = function( id ) {
 
 // We've receive a full object from the network. Update the local instance and
 // notify of the new object version
-internal.updateObjectVersion = function( id, version, data, original, patch, acknowledged ) {
-	var notify,
-		changes,
-		change,
-		patch,
-		localModifications,
-		remoteModifications,
-		transformed,
-		update;
-	// If it's not an ack, it's a change initiated on a different client
+internal.updateObjectVersion = function( entityId, version, entityData, original, incomingPatch, localChange ) {
+	// if the change originated locally then all we need to do
+	// is update the ghost copy and acknowledge the update
+	if (localChange) {
+		return this
+			.store
+			.put( entityId, version, entityData )
+			.then( internal.updateAcknowledged.bind( this, localChange ) );
+	}
+
+	// If we don't have a record of the change already it means that
+	// it originated in another remote client
 	// we need to provide a way for the current client to respond to
 	// a potential conflict if it has modifications that have not been synced
-	if ( !acknowledged ) {
-		changes = this.localQueue.dequeueChangesFor( id );
-		localModifications = change_util.compressChanges( changes, original );
-		remoteModifications = patch;
-		transformed = change_util.transform( localModifications, remoteModifications, original );
-		update = data;
-
-		// apply the transformed patch and emit the update
-		if ( transformed ) {
-			patch = transformed;
-			update = change_util.apply( transformed, data );
-			// queue up the new change
-			change = change_util.modify( id, version, patch );
-			this.localQueue.queue( change );
-		}
+	const locallyQueuedChanges = this.localQueue.dequeueChangesFor( entityId );
+	const localModifications = change_util.compressChanges( locallyQueuedChanges, original );
+	const patchWithLocalChanges = change_util.transform( localModifications, incomingPatch, original );
 
-		notify = this.emit.bind( this, 'update', id, update, original, patch, this.isIndexing );
-	} else {
-		notify = internal.updateAcknowledged.bind( this, acknowledged );
+	const updatedData = patchWithLocalChanges
+		? change_util.apply( patchWithLocalChanges, entityData )
+		: entityData;
+
+	// TODO: Will this ever be false at this point?
```

Comment:
It looks like `change_util.transform` should always return an object, which is always truthy, so I'm wondering if we need these conditionals at this point or should simplify and stick with the straight execution flow

could we ever get a falsy value from `transform()`?

Reply:
in 740013a I reasoned that indeed this condition will always be true because `change_util.transform` will always return at least an empty object and therefore will be truthty, so I removed the condition entirely